pkg/etcdcli/etcdcli_pool_test.go

```diff
@@ -0,0 +1,430 @@
+	assert.Equal(t, 2, poolRecorder.numEndpointCalls)
+	assert.Equal(t, 2, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+}
+
+func TestClientReusesClientsReturned(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+
+	client, err := poolRecorder.pool.Get()
+	require.NoError(t, err)
+	assert.NotNil(t, client)
+	assert.Equal(t, 1, poolRecorder.numNewCalls)
+	assert.Equal(t, 1, poolRecorder.numEndpointCalls)
+	assert.Equal(t, 1, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+	poolRecorder.pool.Return(client)
+	client, err = poolRecorder.pool.Get()
+	require.NoError(t, err)
+	assert.NotNil(t, client)
+	assert.Equal(t, 1, poolRecorder.numNewCalls)
+	assert.Equal(t, 2, poolRecorder.numEndpointCalls)
+	assert.Equal(t, 2, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+}
+
+func TestClientClosesOnChannelCapacity(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+	var clients []*clientv3.Client
+	for i := 0; i < maxNumCachedClients+1; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	// returning all should make sure the last one tripping over capacity should get closed
+	for _, client := range clients {
+		poolRecorder.pool.Return(client)
+	}
+
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numHealthCalls)
+	assert.Equal(t, 1, poolRecorder.numCloseCalls)
+}
+
+func TestNewClientWithTickets(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+	var clients []*clientv3.Client
+	for i := 0; i < maxNumClientTickets; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+
+	// this should block and return an error
+	client, err := poolRecorder.pool.Get()
+	assert.Nil(t, client)
+	assert.Errorf(t, err, "too many active cache clients, rejecting to create new one")
+	// returning one should unlock the get again
+	poolRecorder.pool.Return(clients[0])
+	client, err = poolRecorder.pool.Get()
```

Comment:
Missing error check.
```suggestion
	client, err = poolRecorder.pool.Get()
	require.NoError(t, err)
```

Reply:
done